### PR TITLE
Add support for replay mods

### PR DESCRIPTION
**Pull request by tgi74**
Adds a way to convert a mod enum to Mod objects, then bind them to the player Mods

### CODE REVIEW

**Comment by Frontear** [comment]
May I ask what this aims to accomplish?

File: osu.Game/OsuGame.cs [hunk 1]
```diff
@@ -194,6 +194,7 @@ protected void LoadScore(Score s)
             }
 
             Beatmap.Value = BeatmapManager.GetWorkingBeatmap(s.Beatmap);
+            Beatmap.Value.Mods.BindTo(new Bindable<IEnumerable<Mod>>(s.Mods));
```
Comment (peppy) [inline osu.Game/OsuGame.cs]: You probably don't want this to be a `BindTo`, but just a set.

File: osu.Game/Rulesets/Ruleset.cs [hunk 2]
```diff
@@ -33,6 +34,8 @@ public IEnumerable<Mod> GetAllMods() => Enum.GetValues(typeof(ModType)).Cast<Mod
 
         public abstract IEnumerable<Mod> GetModsFor(ModType type);
 
+        public abstract IEnumerable<Mod> GetLegacyModsFor(LegacyMods mods);
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Ruleset.cs]: * Virtual, not abstract.
* Rename to `ConvertLegacyMods`.
* xmldoc something like `Converts mods from legacy enum values. Do not override if you're not a legacy ruleset.`

File: osu.Game/Beatmaps/Legacy/LegacyMods.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        FadeIn = 1048576,
+        Random = 2097152,
+        Cinema = 4194304,
+        TargetPractice = 8388608,
+        Key9 = 16777216,
+        Coop = 33554432,
+        Key1 = 67108864,
+        Key3 = 134217728,
+        Key2 = 268435456,
+    }
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/Legacy/LegacyMods.cs]: Use the following instead:
```
[Flags]
public enum Mods
{
    None                   = 0,
    NoFail                 = 1 << 0,
    Easy                   = 1 << 1,
    TouchDevice            = 1 << 2,
    Hidden                 = 1 << 3,
    HardRock               = 1 << 4,
    SuddenDeath            = 1 << 5,
    DoubleTime             = 1 << 6,
    Relax                  = 1 << 7,
    HalfTime               = 1 << 8,
    Nightcore              = 1 << 9,
    Flashlight             = 1 << 10,
    Autoplay               = 1 << 11,
    SpunOut                = 1 << 12,
    Relax2                 = 1 << 13,
    Perfect                = 1 << 14,
    Key4                   = 1 << 15,
    Key5                   = 1 << 16,
    Key6                   = 1 << 17,
    Key7                   = 1 << 18,
    Key8                   = 1 << 19,
    FadeIn                 = 1 << 20,
    Random                 = 1 << 21,
    Cinema                 = 1 << 22,
    Target                 = 1 << 23,
    Key9                   = 1 << 24,
    KeyCoop                = 1 << 25,
    Key1                   = 1 << 26,
    Key3                   = 1 << 27,
    Key2                   = 1 << 28,
}
```

File: osu.Game.Rulesets.Catch/CatchRuleset.cs [hunk 4]
```diff
@@ -29,6 +30,45 @@ public class CatchRuleset : Ruleset
             new KeyBinding(InputKey.Shift, CatchAction.Dash),
         };
 
+        public override IEnumerable<Mod> GetLegacyModsFor(LegacyMods mods)
+        {
+            if (mods.HasFlag(LegacyMods.NightCore))
+                yield return new CatchModNightcore();
+            else if (mods.HasFlag(LegacyMods.DoubleTime))
+                yield return new CatchModDoubleTime();
+
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Catch/CatchRuleset.cs]: Extra newline.

File: osu.Game.Rulesets.Osu/OsuRuleset.cs [hunk 5]
```diff
@@ -66,6 +67,54 @@ public override IEnumerable<BeatmapStatistic> GetBeatmapStatistics(WorkingBeatma
             };
         }
 
+        public override IEnumerable<Mod> GetLegacyModsFor(LegacyMods mods)
+        {
+            if (mods.HasFlag(LegacyMods.NightCore))
+                yield return new OsuModNightcore();
+            else if (mods.HasFlag(LegacyMods.DoubleTime))
+                yield return new OsuModDoubleTime();
+
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/OsuRuleset.cs]: Extra newline.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]

**Comment by smoogipoo** [comment]
Looks good to me 👍

**Pull request merged**